superbench/benchmarks/micro_benchmarks/computation_communication_overlap.py

```diff
@@ -0,0 +1,242 @@
+            torch.cuda.set_device(self.__local_rank)
+
+        return True
+
+    def __kernel_nccl_pipeline(self, kernel, matA, matB, stages, message, times):
+        """Computation and nccl kernel pipeline with single GPU.
+
+        Args:
+            kernel (ComputationKernelType): the type of the computation kernel to run
+            matA (list[tensor]): the matrix list used in matmul or mul for every stage
+            matB (tensor): the matrix used in matmul
+            stages (int): the ratio number of computation kernel and communication kernel
+            message(tensor): the data used to be transferred through NCCL
+            times(int): number of times in one step to run
+
+        Return:
+            True of False: if computation kernel type is invalid, return False, else, return True
+        """
+        for i in range(times):
+            if self.__world_size > 1:
+                torch.distributed.all_reduce(message, op=torch.distributed.ReduceOp.SUM, async_op=True)
+            for stage in range(stages):
+                if kernel == ComputationKernelType.MUL:
+                    matA[stage].mul(matA[stage])
+                elif kernel == ComputationKernelType.MATMUL:
+                    matA[stage].matmul(matB)
+                else:
+                    logger.error(
+                        'Unknown comoputation kernel type - benchmark: {}, type: {}.'.format(self._name, kernel)
+                    )
+                    return False
+        return True
+
+    def _benchmark(self):
+        """Implementation for benchmarking."""
+        M = self._args.m
+        K = self._args.k
+        N = self._args.n
+        P = self._args.p
+        Q = self._args.q
+        kernel = self._args.kernel
+        if self.__local_rank == 0:
+            logger.info(
+                'Computation Communication Overlap - using {} GPUs,\
+                matrix shape for computation: M={} K={} N={},\
+                message tensor shape of nccl = [{},{}],\
+                ratio between computation kernel and nccl kernel={}'.format(
+                    self.__world_size, M, K, N, P, Q, self._args.ratio
+                )
+            )
+
+        MatA = list()
+        MatB = list()
+        # Matrix A
+        for _ in range(self._args.ratio):
+            MatA.append(torch.randn(M, K).cuda())
+        # Matrix B
+        MatB = torch.randn(K, N).cuda()
+        # message for nccl to transport
+        shape = [P, Q]
+        message = torch.randn(*shape).cuda()
+
+        # warm up
+        for i in range(self._args.num_warmup):
+            if not self.__kernel_nccl_pipeline(kernel, MatA, MatB, self._args.ratio, message, times=100):
+                return False
+        torch.cuda.synchronize()
+
+        # run and collect results
+        start = time.perf_counter()
+        for i in range(self._args.num_steps):
+            self.__kernel_nccl_pipeline(kernel, MatA, MatB, self._args.ratio, message, times=100)
```

Comment:
Really strange because NCCL is also launched in this function, but we would like to know the computation kernel time in this function. Why do we measure the completion of `self.__kernel_nccl_pipeline()`?

Reply:
I mean to measure the total duration from computation start to computation end，when the self.__kernel_nccl_pipeline() complete, it means all the computation have completed because they are not async. I think we can't get computation kernel time directly if we don't use profiling tools like nsight.